### ISSUE TITLE
Pre-trained Entity Extractor for Foreign Languages

### ISSUE THREAD

**Issue by robinsongh381** [RasaHQ/rasa_nlu#1753]
**Rasa NLU version**:  0.13.8

**Operating system** (windows, osx, ...): Ubuntu 16.04

**Content of model configuration file**:
```yml
language: "kr"
pipeline:
- name: "component.KoreanTokenizer"
- name: "component.PreTrainedCRF"
- name: "component.DomainSpecificCRF"
- name: "ner_synonyms"
- name: "intent_featurizer_count_vectors"
- name: "intent_classifier_tensorflow_embedding"
- name: "intent_entity_featurizer_regex"

```

**Idea**:

Me and my colleagues are currently developing a dialogue system for Korean.

- I am trying to adopt a pre-trained NER component (component.PreTrainedCRF) which can extract Name, Place, Organization, Time and Date, just like ner_duckling for English.

- My plan is to pass user input to component.PreTrainedCRF where general entities (Name, Placec, ..etc) are extracted first and then the same user input is passed to the second CRF model (component.DomainSpecificCRF_ whee domain-dependent entities are extracted (eg. cuisine_type)

**Issues**:

- I have trained component.PreTrainedCRF based on a large corpus , producing "pre_trained_crf_model.pkl"

- However, I cannot find any document that describes how to use the pkl file for further use. I have read [https://medium.com/rasa-blog/enhancing-rasa-nlu-models-with-custom-components-6f54040c4a77] this article which is about adding a custom component, but my case is different that I would like to add another CRF model (component.PreTrainedCRF)

Please let me know how to bridge the two CRF models

I would like to emphasize that I have two different training data sets : one is the large corpus which is for training component.PreTrainedCRF, and the other is "usual" training md file which is for training component.DomainSpecificCRF

Please note that #822 was not helpful for this issue

**Comment by akelad** [comment]
Hi @robinsongh381 really cool that you're building a bot in Korean! Would you be willing to share your custom tokenizer for Korean (e.g. on the forum)? Even the pretrained one would be interesting to see at some point when it's ready.

As for your issue: 
You would just load this pkl file at run time again, just like it's done for the current CRF pkl file. Take a look at the function here: https://github.com/RasaHQ/rasa_nlu/blob/master/rasa_nlu/extractors/crf_entity_extractor.py#L344

Then at runtime, when you need to predict entities, the function should be similar to here: https://github.com/RasaHQ/rasa_nlu/blob/master/rasa_nlu/extractors/crf_entity_extractor.py#L151

Basically, you can take a look at all the classes in this folder and get some inspiration from there :) https://github.com/RasaHQ/rasa_nlu/tree/master/rasa_nlu/extractors

Let me know if that helps!

**Comment by robinsongh381** [comment]
Thanks for the tips !

I will discuss the sharing of the tokenizer component with my colleagues who made the component, and i will give you a response soon !

Cheers

**Comment by robinsongh381** [comment]
Hello @akelad
Having tried your advice, things went well during nlu training BUT I had an error while I was running for evaluation.
  Traceback (most recent call last):
  File "/usr/lib/python3.5/runpy.py", line 184, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.5/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/gunsu/rasa/lib/python3.5/site-packages/rasa_nlu/evaluate.py", line 857, in <module>
    main()
  File "/home/gunsu/rasa/lib/python3.5/site-packages/rasa_nlu/evaluate.py", line 851, in main
    cmdline_args.histogram)
  File "/home/gunsu/rasa/lib/python3.5/site-packages/rasa_nlu/evaluate.py", line 627, in run_evaluation
    interpreter = Interpreter.load(model_path, component_builder)
  File "/home/gunsu/rasa/lib/python3.5/site-packages/rasa_nlu/model.py", line 293, in load
    skip_validation)
  File "/home/gunsu/rasa/lib/python3.5/site-packages/rasa_nlu/model.py", line 320, in create
    model_metadata, **context)
  File "/home/gunsu/rasa/lib/python3.5/site-packages/rasa_nlu/components.py", line 419, in load_component
    cached_component, **context)
  File "/home/gunsu/rasa/lib/python3.5/site-packages/rasa_nlu/registry.py", line 134, in load_component_by_name
    return component_clz.load(model_dir, metadata, cached_component, **kwargs)
  File "/home/gunsu/Desktop/Korean_starter-pack-rasa-stack-master/component/pre_trained_crf.py", line 57, in load
    return cls(meta, ent_tagger)
TypeError: __init__() takes from 1 to 2 positional arguments but 3 were given


Below is my component.PreTrainedCRF

    from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals
import logging
import os
from konlpy.tag import Mecab
import eli5
import typing
from builtins import str
from typing import Any, Dict, List, Optional, Text, Tuple

from rasa_nlu.config import RasaNLUModelConfig, InvalidConfigError
from rasa_nlu.extractors import EntityExtractor
from rasa_nlu.model import Metadata
from rasa_nlu.training_data import Message
from rasa_nlu.training_data import TrainingData

# try:
#     import spacy
# except ImportError:
#     spacy = None

logger = logging.getLogger(__name__)


CRF_MODEL_FILE_NAME = "pre_trained_crf_model.pkl"


class PreTrainedTest(EntityExtractor):

    def process(self, message: Message, **kwargs: Any) -> None:

        extracted = self.add_extractor_name(self.extract_entities(message))
        message.set("entities", message.get("entities", []) + extracted,
                    add_to_output=True)

    @classmethod
    def load(cls,
             meta: Dict[Text, Any],
             model_dir: Text = None,
             model_metadata: Metadata = None,
             cached_component: Optional['PreTrainedTest'] = None,
             **kwargs: Any
             ) -> 'PreTrainedTest':

        from sklearn.externals import joblib

        #file_name = meta.get("file")
        file_name = "pre_trained_crf_model.pkl"
        model_dir = "models/current/nlu"

        model_file = os.path.join(model_dir, file_name)

        if os.path.exists(model_file):
            ent_tagger = joblib.load(model_file)
            return cls(meta, ent_tagger)
        else:
            return cls(meta)

**Comment by robinsongh381** [comment]
@akelad Any comments would be appreciated

**Comment by akelad** [comment]
`TypeError: init() takes from 1 to 2 positional arguments but 3 were given` you should define an init method for your class here

**Comment by robinsongh381** [comment]
Having added the init method, I still have the same error.
What went wrong ?

Thanks

    class PreTrainedCRF(EntityExtractor):

    name = "component.PreTrainedCRF"

    provides = ["entities"]

    requires = []

    defaults = {}

    def __init__(self, component_config=None):

        super(PreTrainedCRF, self).__init__(component_config)

    def process(self, message: Message, **kwargs: Any) -> None:

        extracted = self.add_extractor_name(self.extract_entities(message))
        message.set("entities", message.get("entities", []) + extracted,
                    add_to_output=True)

    @classmethod
    def load(cls,
             meta: Dict[Text, Any],
             model_dir: Text = None,
             model_metadata: Metadata = None,
             cached_component: Optional['PreTrainedCRF'] = None,
             **kwargs: Any
             ) -> 'PreTrainedCRF':

        from sklearn.externals import joblib

        #file_name = meta.get("file")
        file_name = "pre_trained_crf_model.pkl"
        model_dir = "models/current/nlu"

        model_file = os.path.join(model_dir, file_name)

        if os.path.exists(model_file):
            ent_tagger = joblib.load(model_file)
            return cls(meta, ent_tagger)
        else:
            return cls(meta)

**Comment by robinsongh381** [comment]
Also, do I have to define what  **extract_entities** method is ?

**Comment by robinsongh381** [comment]
I have made few changes to PreTrainedCRF and now facing another error -

        extracted = self.add_extractor_name(self.extract_entities(message))
AttributeError: 'PreTrainedCRF' object has no attribute 'extract_entities'

Which I had expected, but I do not want to add existing **extract_entities** method from
https://github.com/RasaHQ/rasa_nlu/blob/1d323eebc0fa6c8f46ff3c2c16af77c1e6870a41/rasa_nlu/extractors/crf_entity_extractor.py#L167
which requires **from_text_to_crf**, **_sentence_to_features** and **_from_crf_to_json** methods, which I have tried and got errors.

Could you describe a simple and neat way of writing **extract_entities** method ?

I have read https://github.com/RasaHQ/rasa_nlu/blob/master/rasa_nlu/extractors/spacy_entity_extractor.py#L33
which I found not that helpful

Cheers

**Comment by akelad** [comment]
@robinsongh381 let's move this to the forum, this is more of a usage question at this point. 
You can change the `extract_entities` method, or use the one from the CRF, whichever works for you
